stan/math/prim/mat/fun/matrix_exp_2x2.hpp

```diff
@@ -0,0 +1,42 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_MATRIX_EXP_2X2_HPP
+#define STAN_MATH_PRIM_MAT_FUN_MATRIX_EXP_2X2_HPP
+
+#include <unsupported/Eigen/MatrixFunctions>
+#include <stan/math/prim/scal/err/check_size_match.hpp>
+
+namespace stan {
+    namespace math {
+
+        using Eigen::Matrix;
+        using Eigen::Dynamic;
+
+        /**
+         * Return the matrix exponential of a 2x2 matrix. Reference for 
+         * algorithm: http://mathworld.wolfram.com/MatrixExponential.html
+         *
+         * @param[in] A A 2x2 matrix
+         * @return Matrix exponential of A.
+         */
+        template <typename T>
+        Matrix<T, Dynamic, Dynamic>
+        matrix_exp_compute_2x2(const Matrix<T, Dynamic, Dynamic>& A) {
+            Matrix<T, Dynamic, Dynamic> B;
+            T a = A(0, 0), b = A(0, 1), c = A(1, 0), d = A(1, 1), delta;
+            delta = sqrt((a - d)*(a - d) + 4*b*c);
+
+            B.resize(2, 2);
+            B(0, 0) = exp(.5*(a+d))
+                       * (delta*cosh(.5*delta) + (a-d)*sinh(.5*delta));
+            B(0, 1) = 2*b*exp(.5*(a+d)) * sinh(.5*delta);
+            B(1, 0) = 2*c*exp(.5*(a+d)) * sinh(.5*delta);
+            B(1, 1) = exp(.5*(a+d))
+                       * (delta*cosh(.5*delta) + (d-a)*sinh(.5*delta));
+
```

Comment:
You want to declare local variables for all the repeated subexpressions, like

```
T half_delta = 0.5 * delta;
T cosh_half_delta = cosh(half_delta);
T sinh_half_delta = sinh(half_delta);
T exp_half_a_plus_d = exp(0.5 * (a + d));
```

and then use those below where you have theme repeated.